File: src/code/cpp/binary_tree/bfs.cpp
#include <queue>

using namespace std;


int fn(TreeNode* root) {
    queue<TreeNode*> que;
    que.push(root);
    int ans = 0;

    while (!que.empty()) {
        int currentLength = que.size();
        // TODO: logic for current level

        for (int i = 0; i < currentLength; i++) {
            TreeNode* node = que.front();
            que.pop();
            // TODO: logic
            if (node->left != nullptr) {
                que.push(node->left);
            }
            if (node->right != nullptr) {
                que.push(node->right);
            }
        }
    }

    return ans;
}

File: src/code/ruby/binary_tree/bfs.rb
def fn(root)
  queue = [root]
  ans   = 0

  until queue.empty?
    current_length = queue.length
    # TODO: logic for current level

    current_length.times do
      node = queue.shift
      # TODO: logic

      queue << node.left if node.left
      queue << node.right if node.right
    end
  end

  ans
end
